### PR TITLE
[Feature | CI] Added a github action to build wheels

### PR DESCRIPTION
**Pull request by Danielkinz**
Changes
--------
Created a new action called "Create Release" (publish.yml) which automatically detects a branch of the format "release/&lt;version&gt;", creates a release with same version, build and uploads wheels for said release.

Supported Versions:
----------------
The following covers which environments the wheels cover:
* Linux with glibc 2.31 or newer (ex. ubuntu 20.04 or newer) - github doesn't have older runners
* Cuda toolkit 11.8 - github runner failed to install torch for cuda 11.7. Older versions not supported by torch >=2.0.0.
* Python 3.8, 3.9, 3.10, 3.11 - torch doesn't support older versions

Notes
-------
* The github action can be edited to fit the main repository's release pattern.
* Example runs and results can be seen in the forked repository

Next Steps
------------
* Replace manual build call with cibuildwheel - could possibly add more flexibility to the wheel versions.
* Add support for windows/macos if needed.
* Add cuda 12.x once torch 2.1.0 becomes stable

### CODE REVIEW

File: .github/workflows/publish.yml [hunk 1]
```diff
@@ -0,0 +1,90 @@
+# This workflow will upload a Python Package to Release asset
+# For more information see: https://help.github.com/en/actions/language-and-framework-guides/using-python-with-github-actions
+
+name: Create Release
+
+on:
+  push:
+    branches:
+      - release/*
```
Comment (WoosukKwon) [inline .github/workflows/publish.yml:9]: A quick question: is there a particular reason for choosing this the trigger? What about making the action triggered whenever a new tag starting with "v" is created? I feel the latter is a bit more straightforward.

File: .github/workflows/publish.yml [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+    name: Build Wheel
+    runs-on: ${{ matrix.os }}
+    needs: release
+    
+    strategy:
+      fail-fast: false
+      matrix:
+          os: ['ubuntu-20.04']
+          python-version: ['3.8', '3.9', '3.10', '3.11']
+          cuda-version: ['11.8'] # Github runner can't build anything older than 11.8
```
Comment (WoosukKwon) [inline .github/workflows/publish.yml:51]: Awesome! BTW, could you provide a link to any document about this restriction?

File: .github/workflows/scripts/env.sh [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+        devscripts \
+        debhelper \
+        nfs-common
+
+# Install python and pip
+# sudo add-apt-repository ppa:deadsnakes/ppa
+# sudo apt install "python$python_version" "python$python_version-distutils" "python$python_version-dev"
+# wget https://bootstrap.pypa.io/get-pip.py
+# python$python_version get-pip.py
+# rm get-pip.py
```
Comment (WoosukKwon) [inline .github/workflows/scripts/env.sh:58]: Did you leave these lines on any purpose? If not, could you delete these?

**Review by WoosukKwon — COMMENTED** [review]
Wonderful! Thank you so much for this amazing work @Danielkinz !
I've checked that the wheels are automatically generated by the GitHub action, and the wheels can be installed very smoothly on python 3.8, 3.9, 3.10, and 3.11. It also works on the NVIDIA PyTorch containers with CUDA 12 installed. 👍

I only request small changes:
1. Could you change the names of the generated wheels from, e.g., `vllm-0.1.3-cp310-cp310-linux_x86_64.whl` to `vllm-0.1.3-cp310-cp310-manylinux1_x86_64.whl` which is compatible with the naming convention in PyPI?
2. When creating a release, could you also include the auto-generated release note?
3. Could you change the GitHub action trigger to any new `v*` tag?

After all, this is very nice work. Thank you so much again @Danielkinz

File: .github/workflows/publish.yml (hunk 1 above)
Comment (Danielkinz) [reply .github/workflows/publish.yml:9]: The main reason for choosing this trigger is since it was easier for me to test with and since I didn't know which trigger you would prefer.
Changing it to trigger on v* tags should not be an issue

File: .github/workflows/publish.yml (hunk 2 above)
Comment (Danielkinz) [reply .github/workflows/publish.yml:51]: I didnt find a document about this restriction.
When I tried using Cuda 11.7 and building the project I received errors about incompatible GPU (even when I tried skipping the runtime error thrown in setup.py).
Might be possible to make 11.7 work, I didn't focus on that too much.

Comment (Danielkinz) [reply .github/workflows/publish.yml:51]: The python versions simply match the ones supported in torch's repo. https://download.pytorch.org/whl/cu118/torch_stable.html

File: .github/workflows/scripts/env.sh (hunk 3 above)
Comment (Danielkinz) [reply .github/workflows/scripts/env.sh:58]: Yes we can delete them. The python installation step didn't work properly at first. This was the first fix, but later I managed to get it to work :P

**Comment by Danielkinz** [comment]
Thank you for the review :)

About the requested changes:
1. Changing the name should pose no issue at all. I already experimented with it a bit. Also, I believe I saw an action to automatically upload to PyPI. Would you like me to check it out as well?
2. How does the autogeneration work currently? I didn't see a .github/release.yml file.
3. Switching to trigger on tag should be simple. If you prefer its also possible to trigger on release to fill in the wheels, which might solve #2 as well

**Comment by WoosukKwon** [comment]
Hi @Danielkinz,

1. Yeah, it would be awesome if the GitHub action can also rename and upload the wheels to PyPI.
2. I think it is possible by using GitHub CLI. Could you refer to [this link](https://stackoverflow.com/questions/75679683/how-can-i-auto-generate-a-release-note-and-create-a-release-using-github-actions)?
3. Yes, please switch the trigger. I personally feel the `v*` tag trigger is a bit more intuitive.

In summary, the ideal workflow for us is that the GitHub action is triggered when a new `v*` tag is created, then it builds python wheels, creates new release with the wheels and an auto-generated release note, and uploads the wheels to PyPI. However, we can skip 2 (release notes) for now if it is not easy.

**Comment by Danielkinz** [comment]
Alright. I'll take a look at it over the next couple of days :)

File: .github/workflows/publish.yml (hunk 2 above)
Comment (WoosukKwon) [reply .github/workflows/publish.yml:51]: Got it. Thanks for the explanation. I think this is totally ok.

File: .github/workflows/scripts/env.sh (hunk 3 above)
Comment (Danielkinz) [reply .github/workflows/scripts/env.sh:58]: At some point I encountered a problem where some wheels failed to upload due to an unknown reason. After some debugging I found out there was a name collision between the wheels. At first I thought there was a problem with the action so I tried to replace it. Upon further investigation I discovered there was a colission between the python that was pre-installed with the new one, and by default it took the pre-installed one. This is the reason some of the scripts define `python_executable=python$1`

Comment (WoosukKwon) [reply .github/workflows/scripts/env.sh:58]: Got it. Thanks for the detailed explanation!

**Comment by WoosukKwon** [comment]
Hi @Danielkinz, what's going on this PR? Please let us know if you need any help.

**Comment by Danielkinz** [comment]
Ah sorry about that. Last night I updated the fork and moved some branches around and the fork's main branch ended up on the same commit as the main repo's.
The requested features are almost done and are currently in testing. The only feature I'm having a bit of trouble with is the PyPI auto upload since its more difficult to test it. (since it requires to set up a test.pypi project)

**Comment by Danielkinz** [comment]
As long as you're fine with it. Theoretically the PyPI auto upload should work, I simply can't test it easily. If you'd like you can try it out on the main branch with test.pypi next time there's a release on the main branch, it shouldn't affect the rest of the pipeline.
I currently encountered another problem - it seems like nvidia has updated their apt repo last night, which causes an error when trying to install their libraries:
"E: Failed to fetch https://developer.download.nvidia.com/compute/cuda/repos/ubuntu2004/x86_64/Packages.gz  File has unexpected size (1127072 != 1126689). Mirror sync in progress? [IP: 152.195.19.142 443]"
Their repo: https://developer.download.nvidia.com/compute/cuda/repos/ubuntu2004/x86_64/
For now I can't continue testing the features since this error kills the pipeline. It should be a temporary problem.

**Comment by WoosukKwon** [comment]
> As long as you're fine with it. Theoretically the PyPI auto upload should work, I simply can't test it easily. If you'd like you can try it out on the main branch with test.pypi next time there's a release on the main branch, it shouldn't affect the rest of the pipeline.

Yes, let's have it in a future PR, as many users are waiting for this PR to be merged. 😃 Could you just let me know if 1) the release notes are now auto-generated, and 2) the wheels are renamed to the format like `vllm-0.1.3-cp310-cp310-manylinux1_x86_64.whl`? I believe these two are also optional and should not block this PR though.

**Comment by Danielkinz** [comment]
Yes, both are implemented. After I test them the PR should be good to go.

**Comment by WoosukKwon** [comment]
@Danielkinz Great, please let me know if the PR is ready.

**Comment by Danielkinz** [comment]
@WoosukKwon The PR is ready :)
Seems like the issue with nvidia's repo was resolved last night so I could test it out.

**Review by WoosukKwon — APPROVED** [review]
@Danielkinz LGTM. Thanks for the super awesome work! This will save many users from the tedious installation issues. Great work!
